### ISSUE TITLE
【POEM】Finished training suddenly and all assert_file_exists output not found

### ISSUE THREAD

**Issue by zp2388** [google-research/google-research#449]
I try to sh 'run.sh', and it seems that I meet all experience requirements, but I find the system Finished training suddenly and all assert_file_exists outputs 'not found'.

 I can't understand what happens, Can someone help me? It's my first time to ask questions in English, this may make my question unclear, and I will try my best to think about anybody's response, thanks!

This is my log:

> Building wheels for collected packages: wrapt, termcolor
  Building wheel for wrapt (setup.py) ... done
  Created wheel for wrapt: filename=wrapt-1.12.1-py3-none-any.whl size=19552 sha256=9bb47c22b3213eae33687d520405a6ad1af358742738563bf5677d2c120cdccf
  Stored in directory: /home/zp/.cache/pip/wheels/4c/8d/0e/ecb228daca7bc2ae7f9a9d713ef75fce4a083de089869418b5
  Building wheel for termcolor (setup.py) ... done
  Created wheel for termcolor: filename=termcolor-1.1.0-py3-none-any.whl size=4830 sha256=747aa37b47651ca7e68b8a56cac12cbf58b0b9ea7bca97f9540358d75237ea5f
  Stored in directory: /home/zp/.cache/pip/wheels/22/6d/f1/cee5814a13ba0c8ecd5ae67093238282c258bfdd6881b6b638
Successfully built wrapt termcolor
Installing collected packages: six, absl-py, numpy, protobuf, zipp, importlib-metadata, markdown, certifi, chardet, urllib3, idna, requests, grpcio, tensorboard-plugin-wit, werkzeug, cachetools, pyasn1, pyasn1-modules, rsa, google-auth, oauthlib, requests-oauthlib, google-auth-oauthlib, tensorboard, h5py, google-pasta, keras-preprocessing, wrapt, gast, opt-einsum, termcolor, tensorflow-estimator, astunparse, tensorflow, decorator, dm-tree, cloudpickle, tensorflow-probability, tf-slim
Successfully installed absl-py-0.11.0 astunparse-1.6.3 cachetools-4.1.1 certifi-2020.6.20 chardet-3.0.4 cloudpickle-1.6.0 decorator-4.4.2 dm-tree-0.1.5 gast-0.3.3 google-auth-1.23.0 google-auth-oauthlib-0.4.1 google-pasta-0.2.0 grpcio-1.33.2 h5py-2.10.0 idna-2.10 importlib-metadata-2.0.0 keras-preprocessing-1.1.2 markdown-3.2.2 numpy-1.18.5 oauthlib-3.1.0 opt-einsum-3.3.0 protobuf-3.13.0 pyasn1-0.4.8 pyasn1-modules-0.2.8 requests-2.24.0 requests-oauthlib-1.3.0 rsa-4.6 six-1.15.0 tensorboard-2.3.0 tensorboard-plugin-wit-1.7.0 tensorflow-2.3.1 tensorflow-estimator-2.3.0 tensorflow-probability-0.11.1 termcolor-1.1.0 tf-slim-1.1.0 urllib3-1.25.11 werkzeug-1.0.1 wrapt-1.12.1 zipp-1.2.0
+ TRAIN_DIR=/tmp/e3d/train
+ mkdir -p /tmp/e3d/train
+ python3 -m poem.train --alsologtostderr --input_table=poem/testdata/tfe-2.tfrecords --train_log_dir=/tmp/e3d/train --batch_size=4 --num_steps=5 --input_shuffle_buffer_size=10 --summarize_percentiles=false
2020-11-03 12:27:16.959669: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcudart.so.10.1
WARNING:tensorflow:From /home/zp/code/pycharm/Pose_Invariant/env3/lib/python3.5/site-packages/tensorflow/python/compat/v2_compat.py:96: disable_resource_variables (from tensorflow.python.ops.variable_scope) is deprecated and will be removed in a future version.
Instructions for updating:
non-resource variables are not supported in the long term
WARNING:tensorflow:From /home/zp/code/pycharm/Pose_Invariant/env3/lib/python3.5/site-packages/tensorflow/python/ops/linalg/linear_operator_full_matrix.py:149: calling LinearOperator.__init__ (from tensorflow.python.ops.linalg.linear_operator) with graph_parents is deprecated and will be removed in a future version.
Instructions for updating:
Do not pass `graph_parents`.  They will  no longer be used.
W1103 12:27:18.325209 140324728178496 deprecation.py:506] From /home/zp/code/pycharm/Pose_Invariant/env3/lib/python3.5/site-packages/tensorflow/python/ops/linalg/linear_operator_full_matrix.py:149: calling LinearOperator.__init__ (from tensorflow.python.ops.linalg.linear_operator) with graph_parents is deprecated and will be removed in a future version.
Instructions for updating:
Do not pass `graph_parents`.  They will  no longer be used.
WARNING:tensorflow:From /home/zp/code/pycharm/Pose_Invariant/poem/core/models.py:97: batch_normalization (from tensorflow.python.keras.legacy_tf_layers.normalization) is deprecated and will be removed in a future version.
Instructions for updating:
Use keras.layers.BatchNormalization instead.  In particular, `tf.control_dependencies(tf.GraphKeys.UPDATE_OPS)` should not be used (consult the `tf.keras.layers.BatchNormalization` documentation).
W1103 12:27:18.362956 140324728178496 deprecation.py:323] From /home/zp/code/pycharm/Pose_Invariant/poem/core/models.py:97: batch_normalization (from tensorflow.python.keras.legacy_tf_layers.normalization) is deprecated and will be removed in a future version.
Instructions for updating:
Use keras.layers.BatchNormalization instead.  In particular, `tf.control_dependencies(tf.GraphKeys.UPDATE_OPS)` should not be used (consult the `tf.keras.layers.BatchNormalization` documentation).
WARNING:tensorflow:From /home/zp/code/pycharm/Pose_Invariant/env3/lib/python3.5/site-packages/tensorflow/python/keras/legacy_tf_layers/normalization.py:336: Layer.apply (from tensorflow.python.keras.engine.base_layer_v1) is deprecated and will be removed in a future version.
Instructions for updating:
Please use `layer.__call__` method instead.
W1103 12:27:18.365832 140324728178496 deprecation.py:323] From /home/zp/code/pycharm/Pose_Invariant/env3/lib/python3.5/site-packages/tensorflow/python/keras/legacy_tf_layers/normalization.py:336: Layer.apply (from tensorflow.python.keras.engine.base_layer_v1) is deprecated and will be removed in a future version.
Instructions for updating:
Please use `layer.__call__` method instead.
WARNING:tensorflow:From /home/zp/code/pycharm/Pose_Invariant/env3/lib/python3.5/site-packages/tensorflow/python/training/moving_averages.py:458: Variable.initialized_value (from tensorflow.python.ops.variables) is deprecated and will be removed in a future version.
Instructions for updating:
Use Variable.read_value. Variables in 2.X are initialized automatically both in eager and graph (inside tf.defun) contexts.
W1103 12:27:18.705586 140324728178496 deprecation.py:323] From /home/zp/code/pycharm/Pose_Invariant/env3/lib/python3.5/site-packages/tensorflow/python/training/moving_averages.py:458: Variable.initialized_value (from tensorflow.python.ops.variables) is deprecated and will be removed in a future version.
Instructions for updating:
Use Variable.read_value. Variables in 2.X are initialized automatically both in eager and graph (inside tf.defun) contexts.
I1103 12:27:18.933850 140324728178496 pipeline_utils.py:308] Resume latest training checkpoint in: /tmp/e3d/train.
WARNING:tensorflow:From /home/zp/code/pycharm/Pose_Invariant/env3/lib/python3.5/site-packages/tensorflow/python/training/adagrad.py:77: calling Constant.__init__ (from tensorflow.python.ops.init_ops) with dtype is deprecated and will be removed in a future version.
Instructions for updating:
Call initializer instance with the dtype argument instead of passing it to the constructor
W1103 12:27:19.484864 140324728178496 deprecation.py:506] From /home/zp/code/pycharm/Pose_Invariant/env3/lib/python3.5/site-packages/tensorflow/python/training/adagrad.py:77: calling Constant.__init__ (from tensorflow.python.ops.init_ops) with dtype is deprecated and will be removed in a future version.
Instructions for updating:
Call initializer instance with the dtype argument instead of passing it to the constructor
WARNING:tensorflow:From /home/zp/code/pycharm/Pose_Invariant/env3/lib/python3.5/site-packages/tf_slim/learning.py:734: Supervisor.__init__ (from tensorflow.python.training.supervisor) is deprecated and will be removed in a future version.
Instructions for updating:
Please switch to tf.train.MonitoredTrainingSession
W1103 12:27:19.663510 140324728178496 deprecation.py:323] From /home/zp/code/pycharm/Pose_Invariant/env3/lib/python3.5/site-packages/tf_slim/learning.py:734: Supervisor.__init__ (from tensorflow.python.training.supervisor) is deprecated and will be removed in a future version.
Instructions for updating:
Please switch to tf.train.MonitoredTrainingSession
2020-11-03 12:27:19.780182: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN)to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2020-11-03 12:27:19.801271: I tensorflow/core/platform/profile_utils/cpu_utils.cc:104] CPU Frequency: 3696000000 Hz
2020-11-03 12:27:19.802094: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x8bb7aa0 initialized for platform Host (this does not guarantee that XLA will be used). Devices:
2020-11-03 12:27:19.802106: I tensorflow/compiler/xla/service/service.cc:176]   StreamExecutor device (0): Host, Default Version
2020-11-03 12:27:19.834374: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcuda.so.1
2020-11-03 12:27:19.927968: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:982] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2020-11-03 12:27:19.928318: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x8bce8b0 initialized for platform CUDA (this does not guarantee that XLA will be used). Devices:
2020-11-03 12:27:19.928330: I tensorflow/compiler/xla/service/service.cc:176]   StreamExecutor device (0): GeForce GTX 1080 Ti, Compute Capability 6.1
2020-11-03 12:27:19.928462: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:982] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2020-11-03 12:27:19.928775: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1716] Found device 0 with properties: 
pciBusID: 0000:01:00.0 name: GeForce GTX 1080 Ti computeCapability: 6.1
coreClock: 1.582GHz coreCount: 28 deviceMemorySize: 10.92GiB deviceMemoryBandwidth: 451.17GiB/s
2020-11-03 12:27:19.928809: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcudart.so.10.1
2020-11-03 12:27:20.214316: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcublas.so.10
2020-11-03 12:27:20.218626: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcufft.so.10
2020-11-03 12:27:20.219405: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcurand.so.10
2020-11-03 12:27:20.224505: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcusolver.so.10
2020-11-03 12:27:20.227276: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcusparse.so.10
2020-11-03 12:27:20.227897: W tensorflow/stream_executor/platform/default/dso_loader.cc:59] Could not load dynamic library 'libcudnn.so.7'; dlerror: libcudnn.so.7: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /usr/local/cuda-10.1/lib64:/usr/local/cuda-10.2/lib64::/usr/local/mpc-0.8.1/lib:/usr/local/gmp-4.3.2/lib:/usr/local/mpfr-2.4.2/lib:/usr/local/gcc-5.3.0/lib:/usr/local/gcc-5.3.0/lib64:/usr/local/cuda-10.1/extras/CUPTI/lib64
2020-11-03 12:27:20.227946: W tensorflow/core/common_runtime/gpu/gpu_device.cc:1753] Cannot dlopen some GPU libraries. Please make sure the missing libraries mentioned above are installed properly if you would like to use GPU. Follow the guide at https://www.tensorflow.org/install/gpu for how to download and setup the required libraries for your platform.
Skipping registering GPU devices...
2020-11-03 12:27:20.227998: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1257] Device interconnect StreamExecutor with strength 1 edge matrix:
2020-11-03 12:27:20.228025: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1263]      0 
2020-11-03 12:27:20.228047: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1276] 0:   N 
INFO:tensorflow:Restoring parameters from /tmp/e3d/train/model.ckpt-00000006
I1103 12:27:20.233745 140324728178496 saver.py:1293] Restoring parameters from /tmp/e3d/train/model.ckpt-00000006
WARNING:tensorflow:From /home/zp/code/pycharm/Pose_Invariant/env3/lib/python3.5/site-packages/tensorflow/python/training/saver.py:1077: get_checkpoint_mtimes (from tensorflow.python.training.checkpoint_management) is deprecated and will be removed in a future version.
Instructions for updating:
Use standard file utilities to get mtimes.
W1103 12:27:20.375499 140324728178496 deprecation.py:323] From /home/zp/code/pycharm/Pose_Invariant/env3/lib/python3.5/site-packages/tensorflow/python/training/saver.py:1077: get_checkpoint_mtimes (from tensorflow.python.training.checkpoint_management) is deprecated and will be removed in a future version.
Instructions for updating:
Use standard file utilities to get mtimes.
INFO:tensorflow:Running local_init_op.
I1103 12:27:20.376783 140324728178496 session_manager.py:505] Running local_init_op.
INFO:tensorflow:Done running local_init_op.
I1103 12:27:20.393647 140324728178496 session_manager.py:508] Done running local_init_op.
INFO:tensorflow:Starting Session.
I1103 12:27:21.270997 140324728178496 learning.py:746] Starting Session.
INFO:tensorflow:Saving checkpoint to path /tmp/e3d/train/model.ckpt
I1103 12:27:21.287984 140318586431232 supervisor.py:1117] Saving checkpoint to path /tmp/e3d/train/model.ckpt
INFO:tensorflow:Starting Queues.
I1103 12:27:21.288069 140324728178496 learning.py:760] Starting Queues.
INFO:tensorflow:Recording summary at step 6.
I1103 12:27:21.492430 140318569645824 supervisor.py:1050] Recording summary at step 6.
INFO:tensorflow:global_step/sec: 0
I1103 12:27:21.823515 140318578038528 supervisor.py:1099] global_step/sec: 0
INFO:tensorflow:Stopping Training.
I1103 12:27:21.906329 140324728178496 learning.py:769] Stopping Training.
INFO:tensorflow:Finished training! Saving model to disk.
I1103 12:27:21.906489 140324728178496 learning.py:777] Finished training! Saving model to disk.
/home/zp/code/pycharm/Pose_Invariant/env3/lib/python3.5/site-packages/tensorflow/python/summary/writer/writer.py:387: UserWarning: Attempting to use a closed FileWriter. The operation will be a noop unless the FileWriter is explicitly reopened.
  warnings.warn("Attempting to use a closed FileWriter. "
+ assert_file_exists /tmp/e3d/train/all_flags.train.json
+ [[ ! -f /tmp/e3d/train/all_flags.train.json ]]
script.sh: 37: script.sh: [[: not found
+ assert_file_exists /tmp/e3d/train/graph.pbtxt
+ [[ ! -f /tmp/e3d/train/graph.pbtxt ]]
script.sh: 37: script.sh: [[: not found
+ assert_file_exists /tmp/e3d/train/model.ckpt-00000005.data-00000-of-00001
+ [[ ! -f /tmp/e3d/train/model.ckpt-00000005.data-00000-of-00001 ]]
script.sh: 37: script.sh: [[: not found
+ assert_file_exists /tmp/e3d/train/model.ckpt-00000005.meta
+ [[ ! -f /tmp/e3d/train/model.ckpt-00000005.meta ]]
script.sh: 37: script.sh: [[: not found
+ assert_file_exists /tmp/e3d/train/model.ckpt-00000005.index
+ [[ ! -f /tmp/e3d/train/model.ckpt-00000005.index ]]
script.sh: 37: script.sh: [[: not found

**Comment by jenjsun** [comment]
Hello, thanks for your interest!
A couple of points:
- The training finishes really fast because run.sh is only for testing the package requirements and checking that the training runs, and not for actual model training. Here train is called with num_steps=5 and so your training finishes quickly. For the paper, we trained with num_steps=5000000 but the model usually converges much earlier, so you could use num_steps=1000000.
- You seem to have a bash syntax error based on the line "script.sh: 37: script.sh: [[: not found". This seem to occur when run.sh is checking for the existence of the checkpoint files. (You can also check this manually by seeing if the files are in /tmp/e3d/train)

**Comment by zp2388** [comment]
Very very thank you!

-now, I change my num_steps and get a good result!

-I have checked /tmp/e3d/train, the checkpoint files is existing.  And  I find the cause of "script.sh: 37: script.sh: [[: not found" is vash syntax, you are right!

Thank you again!

**Comment by jenjsun** [comment]
No prob, glad it worked out! Feel free to close this issue if everything is good. :)
